### PR TITLE
Allow determining exclusive resources programmatically

### PR DESCRIPTION
**Pull request by VladimirDmitrienko**
Solves: #2677

## Overview

Introduce adding `shared resources` programmatically via `@ResourceLock(providers = ...)` attribute
which accepts an array of one or more classes implementing `ResourceLocksProvider` interface.

It allows to add resources in runtime (immediately before starting execution on the engine level).

Resources can be distributed based on any test class / nested test class / test method attributes (e.g. package name, class / method name etc.) or any other user logic.

This approach serves as a more flexible and less verbose alternative for cases in which:

- adding lots of `@ResourceLock(value, mode)` annotations manually may be inconvenient;
- shared resources are not known until runtime.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/)
- [x] Change is documented in the [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksFrom.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * @see ResourceAccessMode
+ * @see ResourceLock
+ * @see ResourceLocksProvider
+ * @see ResourceLocksProvider.Lock
+ */
+@API(status = EXPERIMENTAL, since = "5.12")
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Inherited
+public @interface ResourceLocksFrom {
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksFrom.java:58]: After discussion in the team, we'd prefer adding a `Class<? extends ResourceLocksProvider>[] providers() default {}` attribute to the existing `@ResourceLock` annotation (and making `value()` optional by giving it an empty default value).

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	 *
+	 * @since 5.12
+	 * @see Isolated
+	 * @see Resources
+	 * @see ResourceAccessMode
+	 * @see ResourceLock
+	 * @see ResourceLocksFrom
+	 * @see ResourceLocksProvider
+	 */
+	final class Lock {
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:121]: I wonder if it we should make this an interface and pass a `LockFactory` to the methods of `ResourceLocksProvider` rather than making the implementation part of the API. Having such a parameter object has proven useful in many cases but it might be overengineering here. Thoughts?

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 3]
```diff
@@ -2930,8 +2930,15 @@ parallel with each other but not while any other test that requires `READ_WRITE`
...
+.Declaring shared resources with `{ResourceLock}` annotation
 ----
-include::{testDir}/example/SharedResourcesDemo.java[tags=user_guide]
+include::{testDir}/example/SharedResourcesWithResourceLockAnnotationDemo.java[tags=user_guide]
+----
+
+[source,java]
+.Declaring shared resources provider with `{ResourceLocksFrom}` annotation
+----
+include::{testDir}/example/SharedResourcesWithResourceLocksFromAnnotationDemo.java[tags=user_guide]
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2941]: We should also mention here (and in the annotations' Javadoc) that "static" and "dynamic" locks are combined.

**Review by marcphilipp — CHANGES_REQUESTED** [review]
Looks very promising! Sorry for the late request to change the API.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java (hunk 2 above)
Comment (VladimirDmitrienko) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:121]: Actually, I considered `Lock` like an immutable data carrier which has no behavior, that's why I'm not sure about an interface in this case.

Also, the current usage of `Lock` reminds `@ResourceLock`:
```java
@ResourceLock("key");
// both create a 'READ_WRITE' resource identified by 'key':
new Lock("key");
```

Maybe you have in mind some use cases for an interface?

Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:121]: Not really, let's go with the record-like data class.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 3 above)
Comment (VladimirDmitrienko) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Updated. 
Mentioned in Javadoc for `@ResourceLock` and `ResourceLocksProvider`.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksFrom.java (hunk 1 above)
Comment (VladimirDmitrienko) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksFrom.java]: Updated.

**Comment by VladimirDmitrienko** [comment]
Hi @marcphilipp 
Maybe you have any updates regarding this PR?

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 4]
```diff
@@ -2927,21 +2927,30 @@ In addition to controlling the execution mode using the `{Execution}` annotation
...
-name can be user-defined or one of the predefined constants in `{Resources}`:
+execution. The shared resource is identified by a unique name which is a `String`.
+The name can be user-defined or one of the predefined constants in `{Resources}`:
 `SYSTEM_PROPERTIES`, `SYSTEM_OUT`, `SYSTEM_ERR`, `LOCALE`, or `TIME_ZONE`.
 
-If the tests in the following example were run in parallel _without_ the use of
-{ResourceLock}, they would be _flaky_. Sometimes they would pass, and at other times they
-would fail due to the inherent race condition of writing and then reading the same JVM
-System Property.
+Since Junit Jupiter 5.12 `{ResourceLock}` annotation has 'providers' attribute
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2934]: Since the User Guide is versioned, we usually avoid mentioning specific versions when features were introduced. Thus, please remove that bit.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 5]
```diff
@@ -2927,21 +2927,30 @@ In addition to controlling the execution mode using the `{Execution}` annotation
...
+Sometimes they would pass, and at other times they would fail due to
+the inherent race condition of writing and then reading the same JVM System Property.
+
+When access to shared resources is declared using the `{ResourceLock}` annotation
+or added with `{ResourceLocksProvider}` interface, the JUnit Jupiter engine uses
+this information to ensure that no conflicting tests are run in parallel.
+This guarantee extends to lifecycle methods of a test class or method.
+For example, if a test method is annotated with a `{ResourceLock}` annotation,
+the "lock" will be acquired before any `@BeforeEach` methods are executed
+and released after all `@AfterEach` methods have been executed.
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2953]: Please revert the changes made to the formatting of these two paragraphs.

File: documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		System.setProperty("my.prop", "banana");
+		assertEquals("banana", System.getProperty("my.prop"));
+	}
+
+	static final class Provider implements ResourceLocksProvider {
+
+		@Override
+		public Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
+			ResourceAccessMode mode;
+			if (testMethod.getName().equals("customPropertyIsNotSetByDefault")) {
```
Comment (marcphilipp) [inline documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java:73]: Maybe it would make for a better example if we'd check for `testMethod.getName().startsWith("canSet")` and invert the `if`?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java [hunk 7]
```diff
@@ -81,7 +87,24 @@ public final Set<TestTag> getTags() {
...
+		// @formatter:off
+		Function<ResourceLocksProvider, Set<ResourceLocksProvider.Lock>> providerToLocks =
+				p -> p.provideForMethod(getTestClass(), getTestMethod());
+
+		Set<ExclusiveResource> fromMethodLevelAnnotation = getExclusiveResourcesFromAnnotation(
+				getTestMethod(),
+				providerToLocks
+		);
+		Stream<ExclusiveResource> fromProvidersInClassLevelAnnotation = getExclusiveResourcesFromAnnotationProviders(
+				getTestClass(),
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java:99]: What about `@Nested` test classes? I think we'll have to collect exclusive resources from providers from the test class and, if it's a `@Nested` one, also from its enclosing class, recursively.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java (hunk 7 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java:99]: On second thought, there is no need to apply locks from the class level to methods as it is redundant. When executing the method, the class-level locks are already held.

Comment (VladimirDmitrienko) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java:99]: > What about `@Nested` test classes? I think we'll have to collect exclusive resources from providers from the test class and, if it's a `@Nested` one, also from its enclosing class, recursively.

I suppose that the behaviour of `ResourceLocksProvider` should be similar to the plain `@ResourceLock("...")`.
Let's consider such an example:
```java
@ResourceLock("A")
static class TestClass {

    @Nested
    @ResourceLock("B")
    class NestedTestClass {
    }
}
```
`NestedTestClass` gets only `B` but **does not** get `A`.

That's why I'm not sure that `@Nested` classes should invoke providers of enclosing classes.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 4 above)
Comment (VladimirDmitrienko) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Got it. Done

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 5 above)
Comment (VladimirDmitrienko) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Done

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java (hunk 7 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java:99]: Please disregard my initial comment. We should neither check the test class nor any of its enclosing classes since that's already being done by the parent test descriptors. Thus, please remove the extra code that sets `fromProvidersInClassLevelAnnotation` etc.

Comment (VladimirDmitrienko) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java:99]: If we remove this code then `LocksProvider#provideForMethod()` will not be invoked for `test()`:
```java
@ResourceLock(providers = LocksProvider.class)
static class TestClass {

    @Test
    void test() {
    }
}
```

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java [hunk 8]
```diff
@@ -180,11 +186,38 @@ public static ExecutionMode toExecutionMode(org.junit.jupiter.api.parallel.Execu
 		throw new JUnitException("Unknown ExecutionMode: " + mode);
 	}
 
-	Set<ExclusiveResource> getExclusiveResourcesFromAnnotation(AnnotatedElement element) {
+	Set<ExclusiveResource> getExclusiveResourcesFromAnnotation(AnnotatedElement element,
+			Function<ResourceLocksProvider, Set<ResourceLocksProvider.Lock>> providerToLocks) {
+		// @formatter:off
+		return Stream.concat(
+				getExclusiveResourcesFromAnnotationValue(element),
+				getExclusiveResourcesFromAnnotationProviders(element, providerToLocks)
```
[the remark this answers is not in the document]
Comment (VladimirDmitrienko) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java]: Done

File: documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java (hunk 6 above)
Comment (VladimirDmitrienko) [reply documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java]: Good point.
Done

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java (hunk 7 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java]: Hmm, I see. I didn't think about that use case. In that case, we also need to ensure `LocksProvider#provideForMethod()` gets called for nested classes, e.g.:

```java
@ResourceLock(providers = LocksProvider.class)
static class TestClass { // A: `LocksProvider#provideForClass()` is called
	@Nested
	class Inner { // B: `LocksProvider#provideForNestedClass()` is called
    	@Test
    	void test() { // C: `LocksProvider#provideForMethod()` is called
    	}
	}
}
```

If I'm reading the current code correctly, neither B nor C would be called.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java (hunk 7 above)
Comment (VladimirDmitrienko) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java]: You're right.
Done

File: platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+				new ExclusiveResource("b2", LockMode.READ)
+		);
+
+		var nestedClassResources = getNestedClassResources(
+				SharedResourcesViaAnnotationValueAndProvidersTestCase.NestedClass.class
+		);
+		assertThat(nestedClassResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("c1", LockMode.READ_WRITE),
+				new ExclusiveResource("c2", LockMode.READ_WRITE)
+		);
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java:156]: The provider declared on the outermost class is called as well, right? I think we should change the test to show that.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java (hunk 9 above)
Comment (VladimirDmitrienko) [reply platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java:156]: Yes, that's right.
But such case is already covered in `ResourceLocksProviderTests#ClassLevelProviderTestCase`.
Do you think we should cover it here as well?

Comment (VladimirDmitrienko) [reply platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java]: Indeed. We didn't test such case.
Added

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 10]
```diff
@@ -2970,8 +2978,15 @@ parallel with each other but not while any other test that requires `READ_WRITE`
 to the same shared resource is running.
 
 [source,java]
+.Declaring shared resources "statically" with `{ResourceLock}` annotation
+----
+include::{testDir}/example/sharedresources/StaticSharedResourcesDemo.java[tags=user_guide]
+----
+
+[source,java]
+.Adding shared resources "dynamically" with `{ResourceLocksProvider}` interface
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2987]:
```suggestion
.Adding shared resources "dynamically" with `{ResourceLocksProvider}` implementation
```

File: documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+			ResourceAccessMode mode;
+			if (testMethod.getName().startsWith("canSet")) {
+				mode = READ_WRITE;
+			}
+			else {
+				mode = READ;
+			}
+			Set<Lock> locks = new HashSet<>();
+			locks.add(new Lock(SYSTEM_PROPERTIES, mode));
+			return locks;
```
Comment (marcphilipp) [inline documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java:81]:
```suggestion
			return Collections.singleton(new Lock(SYSTEM_PROPERTIES, mode));
```

File: documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+		@Override
+		public Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
+			ResourceAccessMode mode;
+			if (testMethod.getName().startsWith("canSet")) {
+				mode = READ_WRITE;
+			}
+			else {
+				mode = READ;
+			}
```
Comment (marcphilipp) [inline documentation/src/test/java/example/sharedresources/DynamicSharedResourcesDemo.java:78]:
```suggestion
			ResourceAccessMode mode = testMethod.getName().startsWith("canSet") ? READ_WRITE : READ;
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLock.java [hunk 13]
```diff
@@ -48,10 +49,20 @@
  * <p>Since JUnit Jupiter 5.4, this annotation is {@linkplain Inherited inherited}
  * within class hierarchies.
  *
+ * <p>Since JUnit Jupiter 5.12, this annotation supports adding shared resources
+ * in runtime via {@link ResourceLock#providers}.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLock.java:53]:
```suggestion
 * dynamically at runtime via {@link ResourceLock#providers}.
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+import org.apiguardian.api.API;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ToStringBuilder;
+
+/**
+ * {@code @ResourceLocksProvider} is used to add shared resources
+ * to a test class and / or its test methods in runtime.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:29]:
```suggestion
 * to a test class and / or its test methods dynamically at runtime.
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+	 */
+	default Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
+		return emptySet();
+	}
+
+	/**
+	 *
+	 * <p>{@link Lock} represents a shared resource.
+	 *
+	 * <p>Each resource is identified by {@link #key}.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:121]:
```suggestion
	 * <p>Each resource is identified by a {@link #key}.
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+	 * In addition,{@link #accessMode} allows you to specify
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:122]:
```suggestion
	 * In addition, the {@link #accessMode} allows to specify
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+	 * In the latter case, the test class or test method
+	 * may be executed concurrently with other tests
+	 * that also require {@code READ} access but not at the same time
+	 * as any other test that requires {@code READ_WRITE} access.
+	 *
+	 * <p>This guarantee extends to lifecycle methods of a test class or method.
+	 * For example, if a {@code Lock} is added to a test method
+	 * then the "lock" will be acquired before any
+	 * {@link BeforeEach @BeforeEach} methods are executed and released after all
+	 * {@link AfterEach @AfterEach} methods have been executed.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java:137]: I don't think we need to repeat all that here, in particular since we link to the Javadoc of `@ResourceLock` in the `@apiNote` below.

```suggestion
```

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java [hunk 18]
```diff
@@ -180,11 +200,46 @@ public static ExecutionMode toExecutionMode(org.junit.jupiter.api.parallel.Execu
...
 
-	Set<ExclusiveResource> getExclusiveResourcesFromAnnotation(AnnotatedElement element) {
+	Set<ExclusiveResource> getExclusiveResourcesFromAnnotations(AnnotatedElement element,
+			Function<ResourceLocksProvider, Set<ResourceLocksProvider.Lock>> providerToLocks) {
+		// @formatter:off
+		List<ResourceLock> ownAnnotations = findRepeatableAnnotations(element, ResourceLock.class);
+		List<ResourceLock> enclosingClassesAnnotations = getEnclosingTestClasses().stream()
+				.map(clazz -> findRepeatableAnnotations(clazz, ResourceLock.class))
+				.flatMap(Collection::stream)
+				.collect(toList());
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java:210]: I think it would be slightly better to get the registered `ResourceLockProviders` from the parent test descriptor rather than collecting them again for every test and reinstantiating them but I'll take care of that in a follow-up PR.

**Review by marcphilipp — APPROVED** [review]
Thanks! I'll take it from here.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java (hunk 18 above)
Comment (VladimirDmitrienko) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java:210]: You're right, many method / nested class descriptors may share the same parent 👍

**Comment by VladimirDmitrienko** [comment]
> Thanks! I'll take it from here.

Thank you! 
This was my first contribution, and it was very interesting and valuable experience.
